autocomplete.go

```diff
@@ -0,0 +1,296 @@
+func suggestLatest(focused string, candidacies [][]string) [][]rune {
+	var suggests [][]rune
+	for _, candidacy := range candidacies {
+		s := getPartsFromSlice(focused, candidacy)
+		if s != nil {
+			suggests = append(suggests, s...)
+		}
+	}
+	removeDuplicates(&suggests)
+
+	return suggests
+}
+
+func containColumns(focused string) bool {
+	_, ok := isContained(focused, []string{
+		"select", // gitql> select [tab
+		"name,",
+		"url,",
+		"push_url,",
+		"owner,",
+		"full_name,",
+		"hash,",
+		"date,",
+		"author,",
+		"author_email,",
+		"committer,",
+		"committer_email,",
+		"message,",
+		"full_message,",
+		"type,",
+	})
+	return ok
+}
+
+// Remove duplicates elements in slice.
+func removeDuplicates(s *[][]rune) {
+	found := make(map[string]bool)
+	j := 0
+	for i, x := range *s {
+		key := string(x)
+		if !found[key] {
+			found[key] = true
+			(*s)[j] = (*s)[i]
+			j++
+		}
+	}
+	*s = (*s)[:j]
+}
+
+func suggestCommands(inputs [][]rune, pos int) [][]rune {
+
+	ln := len(inputs)
+
+	if ln == 1 {
+		// When nothing is input yet
+		return [][]rune{[]rune("select")}
+	}
+	focused := string(inputs[ln-2])
+	focus := inputs[ln-1]
+	if focused == "select" {
+		// gitql> select [tab
+		// In the case where the most recent input is "select"
+		return [][]rune{
+			[]rune("*"),
+			[]rune("name"),
+			[]rune("url"),
+			[]rune("push_url"),
+			[]rune("owner"),
+			[]rune("full_name"),
+			[]rune("hash"),
+			[]rune("date"),
+			[]rune("author"),
+			[]rune("author_email"),
+			[]rune("committer"),
+			[]rune("committer_email"),
+			[]rune("message"),
+			[]rune("full_message"),
+			[]rune("type"),
+		}
+	} else if containColumns(focused) {
+		if pos > 0 {
+			// gitql> select na[tab
+			// gitql> select commi[tab
+			// In the case is inputting column
+			return suggestColumnsFromInputting(focus, pos)
+		}
+
+		// gitql> select name, [tab
+		// gitql> select committer, [tab
+		// In the case where the most recent input is the column name and comma
+		return suggestColumnsFromLatest(focused)
+	} else if focused == "from" {
+		// gitql> select * from re[tab
+		// gitql> select * from bran[tab
+		// In the case is inputting table name
+		if pos > 0 {
+			return suggestTablesFromInputting(focus, pos)
+		}
+
+		// gitql> select * from [tab
+		// In the case after inputted "from"
+		return [][]rune{
+			[]rune("remotes"),
+			[]rune("tags"),
+			[]rune("branches"),
+			[]rune("commits"),
+			[]rune("refs"),
+		}
+	} else if focused == "order" {
+		return [][]rune{[]rune("by")}
+	} else if focused == "where" || focused == "by" || focused == "or" || focused == "and" {
+		if pos > 0 {
+			// gitql> select name from remotes where na[tab
+			// gitql> select * from commits where committer = "K" order by com[tab
+			// gitql> select * from commits where committer = "K" or com[tab
+			// In the case is inputting column inputted after "where", "by", "and", "or"
+			return suggestColumnsFromInputting(focus, pos)
```

Comment:
I tried to type
`select message from commits where`  tab  or 
`select mesage from commits where date > '2017-01-01 and ' tab 

And I saw the suggestions for reserved word instead of columns. It would be nice put unit tests for the functions, I can help you.